### ISSUE TITLE
Socket timeout error 

### ISSUE THREAD

**Issue by Tunguska55** [googleapis/google-api-python-client#1080]
I've run into an issue that I can't seem to work around:

> Traceback (most recent call last):
>   File "<https://sysjenkins.b.com/job/GSync/job/google.group.sync/ws/gas/gas.py",> line 535, in <module>
>     main(sys.argv)
>   File "<https://sysjenkins.b.com/job/GSync/job/google.group.sync/ws/gas/gas.py",> line 425, in main
>     add_google_user = googmgmt.add_member(admin, email=user['email'], google_group=row['google_name'])
>   File "<https://sysjenkins.b.com/job/GSync/job/google.group.sync/ws/gas/entutil/googmgmt.py",> line 135, in add_member
>     ggroup = admin.members().insert(groupKey=key, body=user_body).execute()
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/_helpers.py", line 134, in positional_wrapper
>     return wrapped(*args, **kwargs)
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/http.py", line 892, in execute
>     headers=self.headers,
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/http.py", line 195, in _retry_request
>     raise exception
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/http.py", line 172, in _retry_request
>     resp, content = http.request(uri, method, *args, **kwargs)
>   File "/dirs/prov/.local/lib/python3.7/site-packages/google_auth_httplib2.py", line 198, in request
>     uri, method, body=body, headers=request_headers, **kwargs)
>   File "/dirs/prov/.local/lib/python3.7/site-packages/httplib2/__init__.py", line 1991, in request
>     cachekey,
>   File "/dirs/prov/.local/lib/python3.7/site-packages/httplib2/__init__.py", line 1651, in _request
>     conn, request_uri, method, body, headers
>   File "/dirs/prov/.local/lib/python3.7/site-packages/httplib2/__init__.py", line 1589, in _conn_request
>     response = conn.getresponse()
>   File "/usr/local/lib/python3.7/http/client.py", line 1321, in getresponse
>     response.begin()
>   File "/usr/local/lib/python3.7/http/client.py", line 296, in begin
>     version, status, reason = self._read_status()
>   File "/usr/local/lib/python3.7/http/client.py", line 257, in _read_status
>     line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
>   File "/usr/local/lib/python3.7/socket.py", line 589, in readinto
>     return self._sock.recv_into(b)
>   File "/usr/local/lib/python3.7/ssl.py", line 1052, in recv_into
>     return self.read(nbytes, buffer)
>   File "/usr/local/lib/python3.7/ssl.py", line 911, in read
>     return self._sslobj.read(len, buffer)
> socket.timeout: The read operation timed out

This issue happens about 6 or 7 minutes into the build and fails right before it attempts to insert a user into a Google Group. Basically, the code grabs a Google Group, compares it to AD for any changes and if it finds any discrepancies, will either remove them from the Google Group or add them to the Google Group. It does this with a predefined list of groups from a sqlite database. 

I assumed this was happening since it's running through a bunch of groups and it's not taking any action up until it reaches this point, which means the connection was becoming stale. However, I have not found a way to close the Google API connection and rebuild it.

I have attempted to:
1) Increase socket timeout -- this doesn't work, it just hangs until it reaches the timeout then reports the same error
2) Move the creation of the discovery variable outside any loop and inside any loop -- error still occurs
3) Add a ton of sleeps to see if it was rate limiting disguised as a timeout -- error still occurs
4) Completely disable IPv6 as a precaution -- still no change

I've fought back and forth with this a bit and can't figure out how to fix this.

I'm using Python 3.7 and here's my subsection of pip freeze:
> google-api-core==1.23.0
> google-api-python-client==1.12.5
> google-auth==1.22.1
> google-auth-httplib2==0.0.4
> google-auth-oauthlib==0.4.2
> googleapis-common-protos==1.52.0
> httplib2==0.18.1

Running on Centos 7.8

If time elapsed is important I can add that to see how long it takes to get to this error messaeg.

Thanks!

**Comment by Tunguska55** [comment]
I would also like to add that this seems to only happen when I update, but not when I read:

> Group requires restriction, currently open
> Traceback (most recent call last):
>   File "gas.py", line 540, in <module>
>     main(sys.argv)
>   File "gas.py", line 405, in main
>     update_perm = googmgmt.set_group_perm_invask(gset, row['google_name'])
>   File "/RepoStorage/google_ad_sync/gas/entutil/googmgmt.py", line 78, in set_group_perm_invask
>     gpu = admin.groups().update(groupUniqueId=gui, body=get_group).execute()
>   File "/usr/local/lib/python3.8/site-packages/googleapiclient/_helpers.py", line 134, in positional_wrapper
>     return wrapped(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/googleapiclient/http.py", line 883, in execute
>     resp, content = _retry_request(
>   File "/usr/local/lib/python3.8/site-packages/googleapiclient/http.py", line 195, in _retry_request
>     raise exception
>   File "/usr/local/lib/python3.8/site-packages/googleapiclient/http.py", line 172, in _retry_request
>     resp, content = http.request(uri, method, *args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/google_auth_httplib2.py", line 197, in request
>     response, content = self.http.request(
>   File "/usr/local/lib/python3.8/site-packages/httplib2/__init__.py", line 1982, in request
>     (response, content) = self._request(
>   File "/usr/local/lib/python3.8/site-packages/httplib2/__init__.py", line 1650, in _request
>     (response, content) = self._conn_request(
>   File "/usr/local/lib/python3.8/site-packages/httplib2/__init__.py", line 1589, in _conn_request
>     response = conn.getresponse()
>   File "/usr/local/lib/python3.8/http/client.py", line 1322, in getresponse
>     response.begin()
>   File "/usr/local/lib/python3.8/http/client.py", line 303, in begin
>     version, status, reason = self._read_status()
>   File "/usr/local/lib/python3.8/http/client.py", line 264, in _read_status
>     line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
>   File "/usr/local/lib/python3.8/socket.py", line 669, in readinto
>     return self._sock.recv_into(b)
>   File "/usr/local/lib/python3.8/ssl.py", line 1241, in recv_into
>     return self.read(nbytes, buffer)
>   File "/usr/local/lib/python3.8/ssl.py", line 1099, in read
>     return self._sslobj.read(len, buffer)
> socket.timeout: The read operation timed out

**Comment by busunkim96** [comment]
@Tunguska55 would you mind sharing a bit of code as well? Does the timeout happen every single time?

**Comment by Tunguska55** [comment]
The timeout happens consistently every time and tends to be at the same place, which makes me think it's hitting some invisible timeout that I can't see? Here's some code and the output:
Note that the main function is just a snippet, if more is needed let me know.
```
Relevant functions that are called (and where the code breaks):
def set_description(admin, newdescription, google_group=None):
    if google_group is None:
        print("No Google Group supplied for description update")
        return None
    getgroup = get_group_info(admin, google_group)
    getgroup['description'] = newdescription
    key = "{}@blueskystudios.com".format(google_group)
    upd = admin.groups().update(groupKey=key, body=getgroup).execute()
    if upd['description'] == newdescription:
        return "Success"
    else:
        return "Failed"
    
def add_member(admin, email, google_group=None):
    if google_group is None:
        print("No Google Group supplied for membership update")
        return None
    if email is None or email == "":
        print("No email supplied for update")
        return None
    userinf = get_user_info(admin, email)
    if userinf is None:
        print("Unable to find user")
        return None
    user_body = {
        "delivery_settings": "ALL_MAIL",
        "email": email, 
        "etag": userinf['etag'], 
        "id": userinf['id'], 
        "kind": "admin#directory#member", 
        "role": "MEMBER", 
        "status": "ACTIVE", 
        "type": "USER"
    }
    key = "{}@blueskystudios.com".format(google_group)
    current = 0
    check_count = 3
    while True:
        try:
            ggroup = admin.members().insert(groupKey=key, body=user_body).execute()
            break
        except HttpError as err:
            if err.resp.status in [409]:
                return "User is already in the group"
            elif err.resp.status in [404]:
                return "Group not found, check spelling or if it exists"
            elif err.resp.status in [403]:
                return "User not found, check spelling"
            elif err.resp.status in [503]:
                print("Rate limit detected, delaying then retrying")
                current += 1
                time.sleep(5)
                if current >= check_count:
                    return "Rate limiting could not be overcome"
            else:
                print("Socket would be caught here?")
                return err
        except Exception as e:
            print("Error caught:{}".format(e))
            print(key)
            print(user_body)
            print(admin)
            return "Failed"
    return "{} has successfully been added to the Google Group".format(email)


Main code:
# For Group management
    admin = googleapiclient.discovery.build('admin', 'directory_v1', credentials=credentials, cache_discovery=False)
    # For Group settings
    gset = googleapiclient.discovery.build('groupssettings', 'v1', credentials=credentials, cache_discovery=False)

    
    c.execute("SELECT * from sync_list ORDER BY google_name")
    syncrows = [dict(row) for row in c.fetchall()]
    
    
    for row in syncrows:
        # I feel there is a chance that the script is hitting rate limits, so I'm going to temper
        # that with a sleep
        time.sleep(5)
        # Separator for ease of viewing
        print("**************************")
        print("Group Name: {}".format(row['google_name']))
        getperm = get_ggroup_perm(gset,row['google_name'])
        # Need to make sure that sync status is on, otherwise disregard entirely
        if row['sync_on'] == "OFF":
            if dryrun:
                print("Sync is turned off (doesn't affect dryrun or reports)")
            elif report:
                pass
            else:
                print("Sync is turned off, disregarding group")
                continue
        # Let's check if group is correctly set, permission wise
        if not report:
            if getperm['whoCanJoin'] == "INVITED_CAN_JOIN" and row['google_name'][-3:] == "_fs":
                print("Group permission is set properly for FS")
            elif getperm['whoCanJoin'] == "CAN_REQUEST_TO_JOIN":
                print("Group permission is set properly")
            else:
                print("Group requires restriction, currently open")
                if dryrun and row['google_name'][-3:] == "_fs":
                    print("CHANGE WhoCanJoin to invite only")
                elif dryrun and row['google_name'][-3:] != "_fs":
                    print("CHANGE WhoCanJoin to ask")
                else:
                    update_perm = set_group_perm_invask(gset, row['google_name'])
                    if update_perm == "Success":
                        print("Group permissions update SUCCESS")
                    else:
                        print("Group permissions update FAILED")
        # Creating a report version where everything is initial sync, despite DB flags
        # this prevents tampering of DB for report purposes
        if not report:
            res = compare_members(conn, admin, row['google_name'], row['ad_name'],row['sync_type'],row['initial_sync'], dryrun=dryrun)
        if report:
            res = compare_members(conn, admin, row['google_name'], row['ad_name'],row['sync_type'],"True", report=report)
        result = []
        if len(res) > 0:
            gdn = admgmt.get_group_dn(conn,row['ad_name'])
            for user in sorted(res, key=lambda x: x['name']):
                if not report and not dryrun:
                    print("Performing actions on {}".format(user['name']))
                if not report:
                    if user['Google_operation'] is not None:
                        if dryrun and row['sync_on'] != "ON":
                            print("Google Group: {} operation on {} -- {}".format(user['Google_operation'],user['name'],user['email']))
                        else:
                            if user['Google_operation'] == "ADD":
                                if row['sync_on'] == "INIT" or row['sync_on'] == "ON":
                                        # For Group management
                                    add_google_user = add_member(admin, email=user['email'], google_group=row['google_name'])
                                    if "successfully" in add_google_user:
                                        result.append(True)
                                    elif add_google_user == "User is already in the group":
                                        print("-- User already in Google Group --")
                                        result.append(True)
                                    else:
                                        result.append(False)
                                    print(add_google_user)
```
Output:

> NOTE: There are about ~12 groups that run before this, I'm just givin the last few where it breaks
> I believe time elapsed is about 7 minutes in
> <user_first> and <user_last> are generic placeholders to maintain user anonymity
> 
> **************************
> Group Name: tls_story_fs
> Group permission is set properly for FS
> Comparing (Google) tls_story_fs to tlsstory (AD)
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> AD Description already up to date
> Google Description already up to date
> **************************
> Group Name: tm_editors_fs
> Group permission is set properly for FS
> Comparing (Google) tm_editors_fs to tmedit (AD)
> Initial sync authorized, ADD only mode activated
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> -- No actions required in Google --
> -- No actions required in AD --
> Performing actions on <user_first> <user_last>
> Traceback (most recent call last):
>   File "<https://sysjenkins.b.com/job/GSync/job/google.group.sync/ws/gas/gas.py",> line 535, in <module>
>     main(sys.argv)
>   File "<https://sysjenkins.b.com/job/GSync/job/google.group.sync/ws/gas/gas.py",> line 425, in main
>     add_google_user = googmgmt.add_member(admin, email=user['email'], google_group=row['google_name'])
>   File "<https://sysjenkins.b.com/job/GSync/job/google.group.sync/ws/gas/entutil/googmgmt.py",> line 135, in add_member
>     ggroup = admin.members().insert(groupKey=key, body=user_body).execute()
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/_helpers.py", line 134, in positional_wrapper
>     return wrapped(*args, **kwargs)
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/http.py", line 892, in execute
>     headers=self.headers,
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/http.py", line 195, in _retry_request
>     raise exception
>   File "/dirs/prov/.local/lib/python3.7/site-packages/googleapiclient/http.py", line 172, in _retry_request
>     resp, content = http.request(uri, method, *args, **kwargs)
>   File "/dirs/prov/.local/lib/python3.7/site-packages/google_auth_httplib2.py", line 198, in request
>     uri, method, body=body, headers=request_headers, **kwargs)
>   File "/dirs/prov/.local/lib/python3.7/site-packages/httplib2/__init__.py", line 1991, in request
>     cachekey,
>   File "/dirs/prov/.local/lib/python3.7/site-packages/httplib2/__init__.py", line 1651, in _request
>     conn, request_uri, method, body, headers
>   File "/dirs/prov/.local/lib/python3.7/site-packages/httplib2/__init__.py", line 1589, in _conn_request
>     response = conn.getresponse()
>   File "/usr/local/lib/python3.7/http/client.py", line 1321, in getresponse
>     response.begin()
>   File "/usr/local/lib/python3.7/http/client.py", line 296, in begin
>     version, status, reason = self._read_status()
>   File "/usr/local/lib/python3.7/http/client.py", line 257, in _read_status
>     line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
>   File "/usr/local/lib/python3.7/socket.py", line 589, in readinto
>     return self._sock.recv_into(b)
>   File "/usr/local/lib/python3.7/ssl.py", line 1052, in recv_into
>     return self.read(nbytes, buffer)
>   File "/usr/local/lib/python3.7/ssl.py", line 911, in read
>     return self._sslobj.read(len, buffer)
> socket.timeout: The read operation timed out

**Comment by busunkim96** [comment]
Hi,

I know you mentioned you tried increasing the timeout, but just to make sure here's the docstring in the library describing how to do that:

https://github.com/googleapis/google-api-python-client/blob/86d8788ee8a766ca6818620f3fd2899be0e44190/googleapiclient/http.py#L1905-L1919

I don't expect that this is a quota error, since as far as I know APIs will explicitly error rather than hang, which seems to be what's happening for you.

I've seen folks use the same client object from `discovery.build()` for much longer than 6-7 minutes without issues so it may be something specific to the environment.

Two suggestions:

(1) Enable httplib2 logging, which will log all the HTTP Traffic.
https://github.com/googleapis/google-api-python-client/blob/8496ebe7e4c282371c831cabdeb390855ff0d270/docs/logging.md#http-traffic
```py
import httplib2
httplib2.debuglevel = 4
```

(2) Try batching requests to see if the result is any different. https://github.com/googleapis/google-api-python-client/blob/master/docs/batch.md

**Comment by Tunguska55** [comment]
Thanks for the response! I enabled http logging, but didn't yield any useful data from it (it just stops where it normally does then hits the timeout with no extra info). However, I have ran a couple of tests with batch requesting and they yield promising results. 

I'm refactoring the logic in my code now to take advantage of batch requesting and will post back when all is done.

**Comment by Tunguska55** [comment]
After mulling over your comment about users having open connections for longer than 6 minutes without issue, I came back to our networking team determined that it was something in our environment. Well...
It turns out that PaloAlto was blocking https://www.googleapis.com via encryption, which killed the connection after a certain period of time. 
Looking further back in my logs I can see I ran into some SSL errors that were self-resolved, but can also ultimately be attributed to PaloAlto encryption.
After PaloAlto decryption, it works perfectly, with no changes to the code.

Sorry to waste your time on this, hopefully this will help someone avoid my issue in the future haha.

**Comment by Megidd** [comment]
For me, it was due to bad internet connection too. I switched from an ADSL Internet connection to a smartphone (access point) Internet connection and the error got resolved.

**Comment by max-torch** [comment]
> For me, it was due to bad internet connection too. I switched from an ADSL Internet connection to a smartphone (access point) Internet connection and the error got resolved.

I encountered a problem similar to this. I tried to get all Google Domain users using the Admin SDK and got a timeout error. When I switched from my home wifi to smartphone 4G mobile data the error got resolved.

I'm worried that this could impact or plague my other web dev work. Does anyone have an idea why this kind of solution works and what could be wrong with the wifi in a scenario like this?
